01_pangram.py

```diff
@@ -0,0 +1,193 @@
+class Pangram:
+  """
+  A pangram is a phrase which contains every letter at least once, such as
+  "the quick brown fox jumps over the lazy dog".
+
+  Write a function which determines if a given string is a pangram.
+
+  This should be an easy problem, so it’s a good excuse to practice Polya’s
+  method. So, please find at least three different strategies for solving
+  this problem, proceeding through Polya’s method each time. In each case,
+  make a note of how you came to that.
+  """
+
+  def is_pangram_1(self, phrase):
+    """
+    Determines whether a given string is a pangram
+    Time analysis references https://wiki.python.org/moin/TimeComplexity
+    where possible
+
+    Overall
+    * time - o(n^2)? Due to the possible nested o(n) calculations in the for loop?
+    * space - o(1) cos everything in the function requires constant space
+    * although if we take storage of the phrase into account then o(n) where n
+    * is length of the phrase
+
+    Function may terminate early if `phrase` is a pangram and all alphabets are
+     reached before the phrase ends
+    """
+    # time - o(1)
+    # space - o(1)
+    # o(26) constant space and time, since the number of ascii lowercase
+    # characters is constant - 26. so o(1) * 26 = o(26)
+    # which simplifies back to o(1)
+    alpha = set('abcdefghijklmnopqrstuvwxyz')
+
+    # overall loop
+    # time - o(n^2)?
+    # * o(n) `for c in phrase`
+    # * within that, another o(n) for `len(alpha)`?
```

Comment:
What's the overall time if `len(alpha)` is o(1)?

Reply:
@robot-dreams if `len(alpha)` is o(1), then overall time is o(n) where n is length of `phrase`. max one iteration through `phrase` in `for c in phrase`